### ISSUE TITLE
Stats fail on searches which returns large numbers of candidates (> 32000)

### ISSUE THREAD

**Issue by camerojo** [talentbeyondboundaries/tbbtalentv2#1160]
Follows on from issue #1054

**Comment by camerojo** [comment]
Copied from issue #1054 ...


Stats on searches returning large numbers of candidates still don't work because of the way we do stats on the results of specific searches (as opposed to stats on the whole database).

See for example Sarah Walders search: https://tctalent.org/admin-portal/search/2976.
Doing stats on that search will fail because the search returns too many candidates.

We hit up against an internal Postgres database limitation described here: https://stackoverflow.com/questions/49274390/postgresql-and-hibernate-java-io-ioexception-tried-to-send-an-out-of-range-inte

Bottom line, without changing the way we implement this, this means a limitation of reporting on searches with results of more than around 32,000 candidates.

I have put a HotFix change into production which avoids the memory overflow error. It also throws an exception which displays to the user on the stats page explaining this limit - and noting that we are working on a solution.

<img width="1305" alt="Screenshot 2024-06-13 at 3 22 04 PM" src="https://github.com/Talent-Catalog/talentcatalog/assets/10580874/8bb47e88-ef30-4c56-8c3f-9f6ece8bb59d">

**Comment by camerojo** [comment]
Note that this bug will also cause searches built on a base search to fail if that base search returns more than 32,000 candidates.

This has the same root cause - namely that both base searches and stats restrict the pool of candidates to be considered by adding a WHERE clause containing something like this:

AND candidate_id in (id1, id2, ..., idn)

Postgres will fail when n > 32,000

We need a different way of limiting the candidates to be considered other than listing all their ids.

For example, using a subquery like...

AND candidate_id IN (SELECT id FROM candidate WHERE ...)

Note that the above solution will only work if the pool of candidates can be be identified by a query on the Postgres database. If the candidates to be considered come from an Elastic search query, that will not work.

If the candidates come from an Elastic search, and there are more that 32,000 of them, we will still be unable to restrict using a WHERE clause. But if we can change the code so that it uses a Postgres subquery where possible that should be a good enough solution for now. In particular it will allow Sarah Walder to run her reports on Lebanon and Jordan.

**Comment by camerojo** [comment]
## Possible implementation

### Saved Search Search
Add to SavedSearchService a method like...

extractPredicateSQL(savedSearch): String

That would extract the SQL representing the filter parameters in the given savedSearch.

That String could be used to generate the associated search query - or it could also be used to limit results of a stats query based on the results of the search.

### CandidateStatsService

Probably should move stats out of the repository - it is messy and has limitations.

Instead construct queries using EntityManager and predicateSQL retrieved as shown above inside a new CandidateStatsService

eg
`
public long getAccountsByPermissionUsingJPQL() throws ParseException {
    Query query = entityManager.createQuery("SELECT COUNT(*) FROM Account a WHERE a.permission = ?1");
    query.setParameter(1, permissionRepository.findByType("admin"));
    return (long) query.getSingleResult();
}
`
from https://www.baeldung.com/spring-data-jpa-row-count

**Comment by camerojo** [comment]
See issue #55 for alternatives to the Criteria API (which we currently use) which make it easier to reuse SQL strings.

(Note that you can get the string from a CriteriaAPI query - but it is messy and not very usable

See https://stackoverflow.com/questions/6276122/can-i-get-the-sql-string-from-a-jpa-query-object)

**Comment by camerojo** [comment]
Note that there a two problems here:

1. Elastic Search is a complete copy of database which may make it slow to copy from Postgres 

Using lists of id is limited

**Comment by camerojo** [comment]
### To Do
Add option to only copy text to es
Do legacy issue to get rid of SearchJoins

- [x] Create test which runs existing SS getting results using existing CriteriaAPI code and also extracted SQL code - assert number of results are the same. This is what will drive the release of this issue.
- [x] Avoid joining language level twice
- [x] New run stats instead of Run new stats
- [x] New run stats needs to use candidate ids if Elastic search
